src/webviews/extension-side/dataviewer/dataViewerDependencyService.node.ts

```diff
@@ -4,132 +4,51 @@
-            await this.installMissingDependencies(interpreter, tokenSource);
-        } finally {
-            tokenSource.dispose();
-        }
+    constructor(
+        @inject(IApplicationShell) applicationShell: IApplicationShell,
+        @inject(IInstaller) installer: IInstaller,
+        @inject(IPythonExecutionFactory) pythonFactory: IPythonExecutionFactory,
+        @inject(IInterpreterService) interpreterService: IInterpreterService,
+        @inject(IsCodeSpace) isCodeSpace: boolean
+    ) {
+        this.withKernel = new KernelDataViewerDependencyImplementation(applicationShell, isCodeSpace);
+        this.withInterpreter = new InterpreterDataViewerDependencyImplementation(
+            applicationShell,
+            installer,
+            pythonFactory,
+            interpreterService,
+            isCodeSpace
+        );
     }
 
-    private async installMissingDependencies(
-        interpreter: PythonEnvironment,
-        tokenSource: CancellationTokenSource
-    ): Promise<void> {
-        sendTelemetryEvent(Telemetry.PythonModuleInstall, undefined, {
-            action: 'displayed',
-            moduleName: ProductNames.get(Product.pandas)!,
-            pythonEnvType: interpreter?.envType
-        });
-        const selection = this.isCodeSpace
-            ? Common.install()
-            : await this.applicationShell.showErrorMessage(
-                  DataScience.pandasRequiredForViewing(),
-                  { modal: true },
-                  Common.install()
-              );
-
-        // All data science dependencies require an interpreter to be passed in
-        // Default to the active interpreter if no interpreter is available
-        const interpreterToInstallDependenciesInto =
-            interpreter || (await this.interpreterService.getActiveInterpreter());
-
-        if (Cancellation.isCanceled(tokenSource.token)) {
-            return;
-        }
-
-        if (selection === Common.install()) {
-            const cancellationPromise = createPromiseFromCancellation({
-                cancelAction: 'resolve',
-                defaultValue: InstallerResponse.Ignore,
-                token: tokenSource.token
-            });
-            // Always pass a cancellation token to `install`, to ensure it waits until the module is installed.
-            const response = await Promise.race([
-                this.installer.install(Product.pandas, interpreterToInstallDependenciesInto, tokenSource),
-                cancellationPromise
-            ]);
-            if (response === InstallerResponse.Installed) {
-                sendTelemetryEvent(Telemetry.UserInstalledPandas);
-            }
+    async checkAndInstallMissingDependencies(executionEnvironment: IKernel | PythonEnvironment): Promise<void> {
+        // IKernel and PythonEnvironment are only types, so I can't compare prototypes or instances of.
+        if (typeof (executionEnvironment as PythonEnvironment).sysPrefix === 'string') {
```

Comment:
I thin a better way to test this is `if ('sysPrefix' in executionEnvironment && executionEnvironment.sysPrefix) {`
This way if we change the arguments tomorrow this line of code will automatically fail to compile (i.e. compilre will catch it) else something like an explicit cast will not be caught and could lead to bugs

Reply:
The closest I can do is `'sysPrefix' in executionEnvironment && typeof executionEnvironment.sysPrefix === 'string'`  (in its own function), to avoid having to set it to something every time we instantiate it in tests.